compiler/mir/mirgen.nim

```diff
@@ -1331,9 +1349,13 @@ proc genVarTuple(c: var TCtx, n: PNode) =
       # generate the assignment:
       c.buildStmt (if isInit: mnkInit else: mnkAsgn):
         genOperand(c, lhs)
-        c.subTree MirNode(kind: mnkPathPos, typ: lhs.typ,
-                          position: i.uint32):
-          c.use val
+        # the temporary tuple is only used for unpacking; it can always be
+        # moved out of. The temporary tuple is not destroyed, so no
+        # destructive move is required
+        c.buildTree mnkMove, lhs.typ:
```

Comment:
Not quite. By way of how the translation works, the field of the temporary is guaranteed to never be used afterwards.

For context, this code is responsible for the translation of `var (a, b) = f()`. Previously, the following MIR was produced by `mirgen`:
```
def_unpack _1 = get()
def a = _1.0
def b = _1.1
```
`def_unpack` is a special kind of definition (read: hack) that makes `_1` visible to the move analyzer, but disables destructor injection.

With the explicit modifiers, `mirgen` can now produce:
```
def_unpack _1 = get()
def a = move _1.0
def b = move _1.1
```
effectively doing the work the move analyzer previously had to do. Since all the parts are guaranteed to be moved out of `_1`, and no control-flow is possible during the unpacking, `_1` doesn't have to be destroyed, and thus a normal move (instead of a destructive move) is fine.

Reply:
Ah, got it. I was thinking of the following, which I should have added to the initial comment, `var someTuple = ...; var (a,b) = someTuple; ...`. here `someTuple` could be used afterwards so I was trying to suss out when/where we do a defensive copy, my thinking was the move analyser because that's when we'd know if the assignment was the last read.

Reply:
Ah, I see. Thank you for clarifying.

> my thinking was the move analyser because that's when we'd know if the assignment was the last read.

Yep, you're thinking thinking is correct. For the `var (a,b) = someTuple` case, `mirgen` produces the following:
```
def_unpack _1 = sink someTuple
...
```
leaving the work of figuring out whether to copy or move the tuple the move analyzer. The relevant part in `mirgen` is [here](https://github.com/nim-works/nimskull/pull/1195/files/faf45f05831e803a723fa0276fb2f0d7bdc52c59#diff-dc0794500e842ec99b3d7e1f3da2cb00fa6ceb7f88f89b428b7df50766f5bca8R1340)